core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp

```diff
@@ -198,12 +206,13 @@ class TeamPolicyInternal<Kokkos::Cuda, Properties...>
   }
 
   inline static int scratch_size_max(int level) {
-    return (
-        level == 0 ? 1024 * 40 :  // 48kB is the max for CUDA, but we need some
-                                  // for team_member.reduce etc.
-            20 * 1024 *
-                1024);  // arbitrarily setting this to 20MB, for a Volta V100
-                        // that would give us about 3.2GB for 2 teams per SM
+    size_t max_shmem = Cuda().cuda_device_prop().sharedMemPerBlock;
+    return (level == 0
+                ? max_shmem - cuda_team_max_reserved_shared_mem
+                :
+                // arbitrarily setting level 1 scratch limit to 20MB, for a
+                // Volta V100 that would give us about 3.2GB for 2 teams per SM
+                20 * 1024 * 1024);
```

Comment:
Consider turning this guy into a local variable too.

Reply:
not sure that its worth it.

Reply:
anyway done